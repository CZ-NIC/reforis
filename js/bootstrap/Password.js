/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


import React from 'react';
import {LABEL_SIZE, FIELD_SIZE} from "./constants";


export default function Password({name, id, label, onChange, value, error, ...props}) {
    return <div className="form-group row">
        <label className={"form-control-label col-sm-" + LABEL_SIZE} htmlFor={id}>{label}</label>
        <div className={"col-sm-" + FIELD_SIZE}>
            <input
                className={"form-control " + (!error ? "" : "is-invalid")}
                type="password"
                name={name}
                id={id}
                onChange={onChange}
                value={value}
                {...props}
            />
            <div className="invalid-feedback">{error}</div>
        </div>
    </div>;
}