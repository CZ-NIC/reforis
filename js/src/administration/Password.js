/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

Password.propTypes = {
    ws: propTypes.object.isRequired
};

export default function Password({ws}) {
    return <>
        <h3>{_('Password settings')}</h3>
    </>
}
