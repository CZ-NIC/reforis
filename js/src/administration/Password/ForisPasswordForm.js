/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';

import PasswordInput from '../../common/bootstrap/PasswordInput';
import CheckBox from '../../common/bootstrap/Checkbox';
import SubmitButton from '../../formContainer/SubmitButton';
import propTypes from 'prop-types';
import {FORM_STATES} from '../../formContainer/hooks';

ForisPasswordForm.propTypes = {
    formData: propTypes.shape(
        {newForisPassword: propTypes.string}
    ).isRequired,
    formState: propTypes.oneOf(
        Object.keys(FORM_STATES).map(key => FORM_STATES[key])
    ).isRequired,
    formErrors: propTypes.shape({}),
    setFormValue: propTypes.func.isRequired,
    postRootPassword: propTypes.func.isRequired,
};

export default function ForisPasswordForm({formData, formState, formErrors, setFormValue, postForisPassword, ...props}) {
    return <form onSubmit={postForisPassword}>
        <h4>{_('Foris password')}</h4>
        <p>{_('Set your password for this administration interface.')}</p>
        <PasswordInput
            withEye={true}
            label={_('New Foris password')}
            value={formData.newForisPassword}
            error={formErrors.newForisPassword}


            onChange={setFormValue(
                value => ({newForisPassword: {$set: value}})
            )}

            {...props}
        />
        <CheckBox
            label={_('Use same password for advanced administration (root)')}
            helpText={_('Same password would be used for accessing this administration interface, for root user in ' +
                'LuCI web interface and for SSH login. Use a strong password! (If you choose not to set the password ' +
                'for advanced configuration here, you will have the option to do so later. Until then, the root account ' +
                'will be blocked.)'
            )}
            checked={formData.sameForRoot}

            onChange={setFormValue(
                value => ({sameForRoot: {$set: value}})
            )}

            {...props}
        />
        <SubmitButton
            state={formState}
            disabled={!!formErrors.newForisPassword}
        />
    </form>
}
