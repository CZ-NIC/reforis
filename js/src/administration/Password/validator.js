/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function validator(formData) {
    const errors = {
        newForisPassword: validatePassword(formData.newForisPassword),
        newRootPassword: !formData.sameForRoot ? validatePassword(formData.newRootPassword) : null,
    };

    if (errors.newForisPassword || errors.newRootPassword)
        return errors;
    return {}
}

function validatePassword(password) {
    if (password === '')
        return _('Password can\'t be empty.');

    if (password.length < 6)
        return _('Password should have min 6 symbols.');

    if (password.length > 128)
        return _('Password should have max 128 symbols.');

    return null
}