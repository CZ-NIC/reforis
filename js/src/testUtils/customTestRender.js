/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
