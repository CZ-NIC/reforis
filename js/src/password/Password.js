/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useCallback, useEffect } from "react";
import PropTypes from "prop-types";

import {
    SUBMIT_BUTTON_STATES, useForm, Spinner, useAPIGet, useAPIPost, API_STATE, useAlert, ALERT_TYPES,
} from "foris";

import API_URLs from "common/API";

import CurrentForisPasswordForm from "./CurrentForisPasswordForm";
import ForisPasswordForm from "./ForisPasswordForm";
import RootPasswordForm from "./RootPasswordForm";

Password.propTypes = {
    postCallback: PropTypes.func,
};

Password.defaultProps = {
    postCallback: () => undefined,
};

export default function Password({ postCallback }) {
    const [formState, onFormChangeHandler, resetFormData] = useForm(validator);

    const resetPasswordForm = useCallback(() => {
        resetFormData({
            currentForisPassword: "",
            newForisPassword: "",
            sameForRoot: false,
            newRootPassword: "",
        });
    }, [resetFormData]);

    useEffect(() => {
        resetPasswordForm();
    }, [resetPasswordForm]);

    const [passwordIsSetState, getPasswordIsSet] = useAPIGet(API_URLs.password);
    useEffect(() => {
        getPasswordIsSet();
    }, [getPasswordIsSet]);

    const [setAlert, dismissAlert] = useAlert();
    const [postState, post] = useAPIPost(API_URLs.password);
    useEffect(() => {
        if (postState.data) {
            if (postState.state === API_STATE.SUCCESS) {
                setAlert(_("Password changed successfully"), ALERT_TYPES.SUCCESS);
                postCallback();
            } else if (postState.state === API_STATE.ERROR) {
                setAlert(postState.data);
            }
            resetPasswordForm();
        }
    }, [postState, resetPasswordForm, postCallback, setAlert]);

    function postForisPassword(event) {
        event.preventDefault();
        dismissAlert();
        const data = {
            foris_current_password: formState.data.currentForisPassword,
            foris_password: formState.data.newForisPassword,
        };
        if (formState.data.sameForRoot) data.root_password = formState.data.newForisPassword;
        post(data);
    }

    function postRootPassword(event) {
        event.preventDefault();
        dismissAlert();
        const data = {
            foris_current_password: formState.data.currentForisPassword,
            root_password: formState.data.newRootPassword,
        };
        post(data);
    }

    if (passwordIsSetState.state === API_STATE.SENDING || !formState.data) {
        return <Spinner className="row justify-content-center" />;
    }

    const isProcessing = postState === API_STATE.SENDING;
    const submitButtonState = isProcessing
        ? SUBMIT_BUTTON_STATES.SAVING
        : SUBMIT_BUTTON_STATES.READY;

    return (
        <>
            <h1>{_("Password")}</h1>
            <h3>{_("Password settings")}</h3>
            {passwordIsSetState.data.password_set && (
                <CurrentForisPasswordForm
                    formData={formState.data}
                    disabled={isProcessing}
                    setFormValue={onFormChangeHandler}
                />
            )}
            <ForisPasswordForm
                formData={formState.data}
                formErrors={formState.errors}
                submitButtonState={submitButtonState}
                disabled={isProcessing}

                setFormValue={onFormChangeHandler}
                postForisPassword={postForisPassword}
            />
            {!formState.data.sameForRoot && (
                <RootPasswordForm
                    formData={formState.data}
                    formErrors={formState.errors}
                    submitButtonState={submitButtonState}
                    disabled={isProcessing}

                    setFormValue={onFormChangeHandler}
                    postRootPassword={postRootPassword}
                />
            )}
        </>
    );
}

function validator(formData) {
    const errors = {
        newForisPassword: validatePassword(formData.newForisPassword),
        newRootPassword: !formData.sameForRoot ? validatePassword(formData.newRootPassword) : null,
    };

    if (errors.newForisPassword || errors.newRootPassword) return errors;
    return {};
}

function validatePassword(password) {
    if (password === "") return _("Password can't be empty.");

    if (password.length < 6) return _("Password should have min 6 symbols.");

    if (password.length > 128) return _("Password should have max 128 symbols.");

    return null;
}
