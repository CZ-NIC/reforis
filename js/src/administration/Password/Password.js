/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useState, useEffect} from 'react';

import {useAPIGetData, useAPIPostData} from '../../forisAPI/hooks';
import {FORM_STATES, useForm} from '../../forisForm/hooks';
import Alert from '../../bootstrap/Alert';
import CurrentForisPasswordForm from './forms/CurrentForisPasswordForm';
import ForisPasswordForm from './forms/ForisPasswordForm';
import RootPasswordForm from './forms/RootPasswordForm';
import validator from './validator';


function usePasswordIsSet() {
    const [isSet, setIsSet] = useState(null);
    const [getData, isReady] = useAPIGetData('password');
    useEffect(() => {
        getData(data => setIsSet(data['password_set']))
    }, []);
    return [isSet, isReady]
}

export default function Password() {
    const [
        formData,
        formErrors,
        setFormData,

        formState,
        setFormState,

        formIsDisabled,
        setFormValue] = useForm(validator);
    const initialFormData = {
        currentForisPassword: '',
        newForisPassword: '',
        sameForRoot: false,
        newRootPassword: '',
    };
    useEffect(() => {
        setFormData(initialFormData);
        setFormState(FORM_STATES.READY);
    }, []);

    const [passwordIsSet, isReady] = usePasswordIsSet();

    const [alert, setAlert] = useState(null);

    function processRequestResult(res) {
        if (res.error) {
            setAlert({type: 'danger', message: res.error});
            setFormData(data => ({...data, currentForisPassword: ''}))
        } else {
            setAlert({
                type: 'success',
                message: _('Password was successfully changed')
            });
            setFormData(initialFormData)
        }
    }

    const postData = useAPIPostData('password');

    function postForisPassword(e) {
        e.preventDefault();
        let data = {
            foris_current_password: formData.currentForisPassword,
            foris_password: formData.newForisPassword
        };
        if (formData.sameForRoot)
            data['root_password'] = formData.newForisPassword;
        postData(data, res => {
            processRequestResult(res)
        })
    }

    function postRootPassword(e) {
        e.preventDefault();
        postData({
            foris_current_password: formData.currentForisPassword,
            root_password: formData.newRootPassword
        }, res => {
            processRequestResult(res)
        })
    }

    function onAlertDismiss() {
        setAlert(null);
    }
    if (!isReady)
        return null;

    return <>
        {alert ? <Alert type={alert.type} message={alert.message} onDismiss={onAlertDismiss}/> : null}
        <h3>{_('Password settings')}</h3>
        {passwordIsSet ?
            <CurrentForisPasswordForm
                formData={formData}
                disabled={formIsDisabled}
                setFormValue={setFormValue}
            />
            : null}
        <ForisPasswordForm
            formData={formData}
            formErrors={formErrors}
            formState={formState}
            disabled={formIsDisabled}

            setFormValue={setFormValue}
            postForisPassword={postForisPassword}
        />
        {!formData.sameForRoot ?
            <RootPasswordForm
                formData={formData}
                formErrors={formErrors}
                formState={formState}
                disabled={formIsDisabled}

                setFormValue={setFormValue}
                postRootPassword={postRootPassword}
            />
            : null}
    </>
}


