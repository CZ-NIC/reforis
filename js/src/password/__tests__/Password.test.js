/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';

import {fireEvent, getByLabelText, getByText, render, wait} from 'customTestRender';
import mockAxios from 'jest-mock-axios';

import Password from '../Password';

describe('<Password/>', () => {
    let passwordContainer;

    beforeEach(async () => {
        const {container} = render(<Password/>);
        mockAxios.mockResponse({data: {password_set: true}});
        await wait(() => getByText(container, 'Advanced administration (root) password'));
        passwordContainer = container;
    });
    it('Snapshot', () => {
        expect(passwordContainer).toMatchSnapshot();
    });
    it('Snapshot: same password for root', () => {
        fireEvent.click(getByLabelText(passwordContainer, 'Use same password for advanced administration (root)'));
        expect(passwordContainer).toMatchSnapshot();
    });
});
