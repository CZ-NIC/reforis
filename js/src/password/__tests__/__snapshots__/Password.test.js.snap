// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Password/> Snapshot 1`] = `
<div>
  <h1>
    Password
  </h1>
  <p>
    Here you can set passwords for Foris and optionally for advanced features (LuCI and SSH). Make sure to set secure password which is long and uncommon.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-3"
  >
    <h2>
      Foris Password
    </h2>
    <p>
      Set your password for this administration interface.
    </p>
    <form>
      <div
        class="form-group"
      >
        <label
          for="1"
        >
          Current Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control"
            id="1"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="2"
        >
          New Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control  is-invalid"
            id="2"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
        <div
          class="invalid-feedback"
        >
          Password can't be empty.
        </div>
      </div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="3"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="3"
          >
            Use same password for advanced administration (root)
            <small
              class="form-text text-muted"
            >
              Same password would be used for accessing this administration interface, for root user in LuCI web interface and for SSH login. Use a strong password! (If you choose not to set the password for advanced configuration here, you will have the option to do so later. Until then, the root account will be blocked.)
            </small>
          </label>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          disabled=""
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
    <form
      class="mt-3"
    >
      <h2>
        Advanced Administration (root) Password
      </h2>
      <p>
        In order to access the advanced configuration options which are not available here, you must
set the root user's password. The advanced configuration options can be managed either 
through the 
        <a
          href="%s"
          rel="noopener noreferrer"
          target="_blank"
        >
          LuCI web interface
        </a>
         or via SSH.
      </p>
      <div
        class="form-group"
      >
        <label
          for="4"
        >
          New advanced administration password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control  is-invalid"
            id="4"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
        <div
          class="invalid-feedback"
        >
          Password can't be empty.
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          disabled=""
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<Password/> Snapshot: same password for root 1`] = `
<div>
  <h1>
    Password
  </h1>
  <p>
    Here you can set passwords for Foris and optionally for advanced features (LuCI and SSH). Make sure to set secure password which is long and uncommon.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-3"
  >
    <h2>
      Foris Password
    </h2>
    <p>
      Set your password for this administration interface.
    </p>
    <form>
      <div
        class="form-group"
      >
        <label
          for="1"
        >
          Current Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control"
            id="1"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="2"
        >
          New Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control  is-invalid"
            id="2"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
        <div
          class="invalid-feedback"
        >
          Password can't be empty.
        </div>
      </div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="3"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="3"
          >
            Use same password for advanced administration (root)
            <small
              class="form-text text-muted"
            >
              Same password would be used for accessing this administration interface, for root user in LuCI web interface and for SSH login. Use a strong password! (If you choose not to set the password for advanced configuration here, you will have the option to do so later. Until then, the root account will be blocked.)
            </small>
          </label>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          disabled=""
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Customized <Password/> Snapshot 1`] = `
<div>
  <h1>
    Password
  </h1>
  <p>
    Here you can set passwords for Foris and optionally for advanced features (LuCI and SSH). Make sure to set secure password which is long and uncommon.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-3"
  >
    <h2>
      Foris Password
    </h2>
    <p>
      Set your password for this administration interface.
    </p>
    <form>
      <div
        class="form-group"
      >
        <label
          for="1"
        >
          Current Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control"
            id="1"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="2"
        >
          New Foris password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control  is-invalid"
            id="2"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="input-group-text"
              type="button"
            >
              <i
                class="fa fa-eye"
              />
            </button>
          </div>
        </div>
        <div
          class="invalid-feedback"
        >
          Password can't be empty.
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          disabled=""
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
